File: python/patterns/producer.py
"""
Reliable Kafka producer patterns.

Key production concerns:
  - Idempotent delivery (acks=all + enable.idempotence=true)
  - Retry with backoff on transient failures
  - Delivery callback for guaranteed confirmation
  - Schema-validated messages
"""
from __future__ import annotations

import json
import time
from dataclasses import dataclass, field
from typing import Any, Callable, Optional


@dataclass
class ProducerConfig:
    bootstrap_servers: str
    topic: str
    acks: str = "all"            # all replicas must ack — no data loss
    retries: int = 5
    retry_backoff_ms: int = 300
    idempotence: bool = True     # exactly-once delivery per partition
    compression: str = "snappy"  # reduce network cost at scale


@dataclass
class Message:
    key: str
    value: dict[str, Any]
    headers: dict[str, str] = field(default_factory=dict)

    def serialize(self) -> bytes:
        return json.dumps(self.value).encode("utf-8")

    def key_bytes(self) -> bytes:
        return self.key.encode("utf-8")


@dataclass
class DeliveryReport:
    topic: str
    partition: int
    offset: int
    latency_ms: float
    error: Optional[str] = None

    @property
    def succeeded(self) -> bool:
        return self.error is None


class ReliableProducer:
    """
    Kafka producer with idempotent delivery and delivery confirmation.

    Production pattern: never fire-and-forget. Always confirm delivery via
    callback before considering a message sent. Track offset for audit trail.

    Usage:
        producer = ReliableProducer(config)
        report = producer.send(Message(key="order-123", value={"amount": 100}))
        assert report.succeeded
    """

    def __init__(self, config: ProducerConfig,
                 delivery_callback: Optional[Callable[[DeliveryReport], None]] = None) -> None:
        self.config = config
        self._delivery_callback = delivery_callback
        self._sent: list[DeliveryReport] = []
        # In production: self._producer = confluent_kafka.Producer({...})

    def send(self, message: Message) -> DeliveryReport:
        """Send one message. Blocks until delivery confirmed."""
        start = time.monotonic()

        # Production:
        # self._producer.produce(
        #     topic=self.config.topic,
        #     key=message.key_bytes(),
        #     value=message.serialize(),
        #     headers=message.headers,
        #     on_delivery=self._on_delivery,
        # )
        # self._producer.flush()  # block until ack

        # Simulated delivery report (replace with real confluent_kafka.Message)
        report = DeliveryReport(
            topic=self.config.topic,
            partition=0,
            offset=len(self._sent),
            latency_ms=(time.monotonic() - start) * 1000,
        )
        self._sent.append(report)
        if self._delivery_callback:
            self._delivery_callback(report)
        return report

    def send_batch(self, messages: list[Message]) -> list[DeliveryReport]:
        """Send a batch. All messages produced before flush — more efficient."""
        reports = []
        for msg in messages:
            reports.append(self.send(msg))
        return reports

    def stats(self) -> dict:
        total = len(self._sent)
        failed = sum(1 for r in self._sent if not r.succeeded)
        return {
            "total": total,
            "succeeded": total - failed,
            "failed": failed,
            "avg_latency_ms": sum(r.latency_ms for r in self._sent) / total if total else 0,
        }

File: go/producer/producer.go
// Package producer implements reliable Kafka producer patterns in Go.
// Uses the confluent-kafka-go client — the same client used by Confluent Cloud.
package producer

import (
	"encoding/json"
	"fmt"
	"time"
)

// Config holds producer configuration.
type Config struct {
	BootstrapServers string
	Topic            string
	Acks             string // "all" for no data loss
	Idempotent       bool   // exactly-once per partition
	CompressionType  string // snappy, lz4, zstd
}

// DefaultConfig returns production-safe defaults.
func DefaultConfig(servers, topic string) Config {
	return Config{
		BootstrapServers: servers,
		Topic:            topic,
		Acks:             "all",
		Idempotent:       true,
		CompressionType:  "snappy",
	}
}

// Message is a Kafka message with typed key and value.
type Message struct {
	Key     string
	Value   map[string]interface{}
	Headers map[string]string
}

// Serialize converts the value to JSON bytes.
func (m Message) Serialize() ([]byte, error) {
	return json.Marshal(m.Value)
}

// DeliveryReport is returned after a message is confirmed delivered.
type DeliveryReport struct {
	Topic     string
	Partition int32
	Offset    int64
	LatencyMs float64
	Err       error
}

func (d DeliveryReport) Succeeded() bool { return d.Err == nil }

// Producer wraps a Kafka producer with delivery confirmation.
type Producer struct {
	config Config
	sent   []DeliveryReport
}

// New creates a new Producer.
// Production: replace mock with confluent_kafka.NewProducer(configMap)
func New(config Config) *Producer {
	return &Producer{config: config}
}

// Send produces one message and blocks until delivery confirmed.
//
// Production implementation:
//
//	p.producer.Produce(&kafka.Message{
//	    TopicPartition: kafka.TopicPartition{Topic: &config.Topic, Partition: kafka.PartitionAny},
//	    Key:   []byte(msg.Key),
//	    Value: serialized,
//	}, deliveryChan)
//	e := <-deliveryChan
func (p *Producer) Send(msg Message) (DeliveryReport, error) {
	start := time.Now()
	serialized, err := msg.Serialize()
	if err != nil {
		return DeliveryReport{Err: err}, fmt.Errorf("serialize: %w", err)
	}
	_ = serialized // production: pass to confluent producer

	report := DeliveryReport{
		Topic:     p.config.Topic,
		Partition: 0,
		Offset:    int64(len(p.sent)),
		LatencyMs: float64(time.Since(start).Milliseconds()),
	}
	p.sent = append(p.sent, report)
	return report, nil
}

// SendBatch produces all messages before flushing — higher throughput than
// sending one-by-one. Production: call p.producer.Flush(30_000) after loop.
func (p *Producer) SendBatch(msgs []Message) ([]DeliveryReport, error) {
	reports := make([]DeliveryReport, 0, len(msgs))
	for _, msg := range msgs {
		r, err := p.Send(msg)
		if err != nil {
			return reports, err
		}
		reports = append(reports, r)
	}
	return reports, nil
}

// Stats returns delivery statistics.
func (p *Producer) Stats() map[string]interface{} {
	failed := 0
	for _, r := range p.sent {
		if !r.Succeeded() {
			failed++
		}
	}
	return map[string]interface{}{
		"total":    len(p.sent),
		"failed":   failed,
		"topic":    p.config.Topic,
	}
}
